File: go/console/serve/topics.go
package serve

import "hop.top/kit/go/runtime/bus"

// DefaultTopicPrefix is the 2-segment source.category prefix serve
// events publish under. Rebrandable by adopters the same way every
// other kit emitter's prefix is (see docs/contracts/event-topics.md).
const DefaultTopicPrefix = "kit.serve"

// Action segments for serve lifecycle events (contract
// §"Surfaced events"). Each satisfies bus.ValidateTopic without
// extending the past-tense whitelist: started/failed/stopped are
// listed there, and ready_reported is a snake_case multi-word action
// that satisfies the "ed" heuristic on the whole segment — the same
// shape as pre_transitioned.
//
// A bare "ready" does not validate. Do not introduce one.
const (
	ActionStarted       = "started"
	ActionReadyReported = "ready_reported"
	ActionFailed        = "failed"
	ActionStopped       = "stopped"
)

// Object segments. The service identifier travels in the event
// payload, not in the topic, so subscribers are not forced to re-bind
// when a tool gains a service.
const (
	ObjectService    = "service"
	ObjectSupervisor = "supervisor"
)

// DefaultTopics returns the conformant topic set for prefix, which is
// a 2-segment source.category string such as DefaultTopicPrefix.
// Keys are "<object>.<action>".
//
// Every returned topic has already passed bus.ValidateTopic by
// construction; a prefix that would produce a non-conformant topic
// makes bus.TopicOf panic at wiring time rather than silently
// dropping events at runtime.
func DefaultTopics(prefix string) bus.TopicMap {
	if prefix == "" {
		prefix = DefaultTopicPrefix
	}
	source, category := splitPrefix(prefix)
	out := bus.TopicMap{}
	for _, spec := range []struct {
		object string
		action string
	}{
		{ObjectService, ActionStarted},
		{ObjectService, ActionReadyReported},
		{ObjectService, ActionFailed},
		{ObjectService, ActionStopped},
		{ObjectSupervisor, ActionReadyReported},
		{ObjectSupervisor, ActionStopped},
	} {
		out[spec.object+"."+spec.action] = bus.TopicOf(source, category, spec.object).Action(spec.action)
	}
	return out
}

// splitPrefix splits a "source.category" prefix. A prefix without a
// dot is taken as the source with the default category, so a
// mis-typed prefix still produces a validatable topic rather than a
// panic deep inside the builder.
func splitPrefix(prefix string) (source, category string) {
	for i := 0; i < len(prefix); i++ {
		if prefix[i] == '.' {
			return prefix[:i], prefix[i+1:]
		}
	}
	return prefix, "serve"
}

File: sdk/experimental/php/src/Serve/Topics.php
<?php

declare(strict_types=1);

namespace HopTop\Kit\Serve;

/**
 * The six lifecycle transitions, as the 4-segment past-tense topic
 * strings the contract fixes.
 *
 * These strings are contract for a port that publishes at all: "a
 * subscriber is written against the string and does not know which
 * language published it". This SDK has no event bus, so nothing here
 * publishes today — the transitions reach an operator through the log
 * sink instead, which the contract explicitly accepts. The vocabulary
 * is still defined here, in one place, because the log carries the
 * same object and action words, and because the day this SDK grows a
 * bus the topics must already be the right strings rather than
 * something invented then.
 */
final class Topics
{
    /** The 2-segment source.category prefix serve events publish under. */
    public const string DEFAULT_PREFIX = 'kit.serve';

    /**
     * Action segments. The action is `ready_reported`, not a bare
     * `ready`: the bare form fails the past-tense validation in
     * event-topics.md, so a port emitting it would publish a topic Go
     * subscribers reject.
     */
    public const string ACTION_STARTED = 'started';
    public const string ACTION_READY_REPORTED = 'ready_reported';
    public const string ACTION_FAILED = 'failed';
    public const string ACTION_STOPPED = 'stopped';

    /**
     * Object segments. The service identifier travels in the payload,
     * not the topic, so subscribers are not forced to re-bind when a
     * tool gains a service.
     */
    public const string OBJECT_SERVICE = 'service';
    public const string OBJECT_SUPERVISOR = 'supervisor';

    /**
     * The conformant topic set for $prefix, keyed `<object>.<action>`.
     *
     * @return array<string, string>
     */
    public static function all(string $prefix = self::DEFAULT_PREFIX): array
    {
        $p = $prefix === '' ? self::DEFAULT_PREFIX : $prefix;
        $out = [];
        foreach (self::pairs() as [$object, $action]) {
            $out["{$object}.{$action}"] = "{$p}.{$object}.{$action}";
        }
        return $out;
    }

    /**
     * The six transitions as object/action pairs, in the order the
     * contract's table lists them.
     *
     * @return list<array{0: string, 1: string}>
     */
    public static function pairs(): array
    {
        return [
            [self::OBJECT_SERVICE, self::ACTION_STARTED],
            [self::OBJECT_SERVICE, self::ACTION_READY_REPORTED],
            [self::OBJECT_SERVICE, self::ACTION_FAILED],
            [self::OBJECT_SERVICE, self::ACTION_STOPPED],
            [self::OBJECT_SUPERVISOR, self::ACTION_READY_REPORTED],
            [self::OBJECT_SUPERVISOR, self::ACTION_STOPPED],
        ];
    }
}
